### PR TITLE
Issue 32 init pretrained weigths

### PR DESCRIPTION
**Pull request by CarolinaFurtado**
Resolving issue #32  - enable training starting from pretrained weights

Tested on tf2.1. Can be tested as e.g:

`python3 train_segmentation_model.py --gcp-bucket gs://necstlab-sandbox --config-file configs/config_sandbox/tf_version_debug//train-small-3class_tfversion_debug.yaml  --random-module-global-seed 1 --numpy-random-global-seed 1 --tf-random-global-seed 1 --pre-trained-model-id segmentation-model-small-3class_tfversion_debug_20201019T174848Z`

If no `--pre-trained-model-id` is given, it is initiated as none and random weights are given. 

The metadata of the pretrained model is also included: 

image

### CODE REVIEW

File: configs/config_sandbox/tf_version_debug/dataset-small-3class_tfversion_debug.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  stack_downsampling:
+    type: 'linear' # 'None', 'random', 'linear', 'from_start', 'from_end'
+#    frac: 1.0  # if 'random', 'linear', 'from_start', or 'from_end' selected; ignored if `None` selected
+    number_of_images: 100  # if 'random', 'linear', 'from_start', or 'from_end' selected; ignored if `None` selected
+    num_skip_beg_slices: 50 # trims n slices off of beginning of stack with N total slices. Slice n+1 becomes new Slice 1
+    num_skip_end_slices: 50 # trims m slices off of end of stack with N total slices. Slice N-(m+1) becomes new last slice
+  target_size: [512, 512]  # width, height
+  image_cropping:
+    type: 'class' # 'None' (downscale to target), 'linear' (def # crops), 'random' (def # crops), `class` (def # crops), 'all' (all crops)
+    num_per_image: None # if 'linear' (translation) or 'random' selected, then is num of crops (of target size) per image
```
Comment (rak5216) [inline configs/config_sandbox/tf_version_debug/dataset-small-3class_tfversion_debug.yaml:21]: should be `Null`

File: configs/config_sandbox/tf_version_debug/train-small-3class_tfversion_debug.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+train_config:
+  model_id_prefix: 'segmentation-model-small-3class_tfversion_debug'
+  dataset_id: 'dataset-small-3class_tfversion_debug'
+  segmentation_model:
+    model_name: 'Unet'
+    model_parameters:
+      backbone_name: 'vgg16'
+      encoder_weights: Null   # Null (random initialization) or 'imagenet' (pre-training on ImageNet)
```
Comment (rak5216) [inline configs/config_sandbox/tf_version_debug/train-small-3class_tfversion_debug.yaml:8]: note that imagenet requires RGB images for user. we'll have to standardize these configs, well update the standard config really after the near-term issues are done

File: requirements.txt [hunk 3]
```diff
@@ -1,5 +1,5 @@
 numpy
-tensorflow-gpu
+tensorflow-gpu==2.1.0
```
Comment (rak5216) [inline requirements.txt:2]: `requirements.txt` should be excluded, ie, not part of PR. the version of tf is case-specific

File: train_segmentation_model.py [hunk 4]
```diff
@@ -191,7 +202,9 @@ def train(gcp_bucket, config_file, random_module_global_seed, numpy_random_globa
         'global_threshold_for_metrics': global_threshold,
         'random-module-global-seed': random_module_global_seed,
         'numpy_random_global_seed': numpy_random_global_seed,
-        'tf_random_global_seed': tf_random_global_seed
+        'tf_random_global_seed': tf_random_global_seed,
+        'pre_trained_model_id': pre_trained_model_id,
+        'pre_trained_model_config': pre_trained_model_config
```
Comment (rak5216) [inline train_segmentation_model.py:207]: we may also want pre trained model metadata, given that it'll contain additional details like seeds. also this should specify train config vs. dataset config

File: train_segmentation_model.py [hunk 5]
```diff
@@ -234,5 +247,10 @@ def train(gcp_bucket, config_file, random_module_global_seed, numpy_random_globa
         type=int,
         default=None,
         help='The setting of tf.random.set_seed(global seed), where global seed is int or default None (no seed given).')
+    argparser.add_argument(
+        '--pre-trained-model-id',
```
Comment (rak5216) [inline train_segmentation_model.py:251]: sry too picky, but i'd simplify slightly to pretrained instead of pre-trained or pre_trained

File: train_segmentation_model.py [hunk 6]
```diff
@@ -67,6 +67,16 @@ def train(gcp_bucket, config_file, random_module_global_seed, numpy_random_globa
     logs_dir = Path(model_dir, 'logs')
     logs_dir.mkdir(parents=True)
 
+    path_pre_trained_model = None
+    pre_trained_model_config = None
```
Comment (rak5216) [inline train_segmentation_model.py:71]: i think put these two lines inside of the `if` statement so it's more readable.
```
path_pre_trained_model = None
pre_trained_model_config = None
```

File: train_segmentation_model.py [hunk 7]
```diff
@@ -67,6 +67,16 @@ def train(gcp_bucket, config_file, random_module_global_seed, numpy_random_globa
...
+    if pre_trained_model_id is not None:
+        local_pre_trained_model_dir = Path(tmp_directory, 'pre_trained_models')
+        copy_folder_locally_if_missing(os.path.join(gcp_bucket, 'models', pre_trained_model_id), local_pre_trained_model_dir)
+        path_pre_trained_model = Path(local_pre_trained_model_dir, pre_trained_model_id, "model.hdf5").as_posix()
+
```
Comment (rak5216) [inline train_segmentation_model.py:76]:
```
else:
    path_pre_trained_model = None
```

File: train_segmentation_model.py [hunk 8]
```diff
@@ -67,6 +67,16 @@ def train(gcp_bucket, config_file, random_module_global_seed, numpy_random_globa
...
+        with Path(local_pre_trained_model_dir, pre_trained_model_id, 'config.yaml').open('r') as f:
+            pre_trained_model_config = yaml.safe_load(f)['train_config']
+
```
Comment (rak5216) [inline train_segmentation_model.py:79]:
```
else:
    path_pre_trained_model = None
```

File: train_segmentation_model.py [hunk 9]
```diff
@@ -67,6 +67,16 @@ def train(gcp_bucket, config_file, random_module_global_seed, numpy_random_globa
     logs_dir = Path(model_dir, 'logs')
     logs_dir.mkdir(parents=True)
 
+    path_pre_trained_model = None
+    pre_trained_model_config = None
+    if pre_trained_model_id is not None:
```
Comment (rak5216) [inline train_segmentation_model.py:72]: @Josh-Joseph, we may want to use `assert`s to enforce continuity of pretrained model config/metadata and current model config, e.g., make sure number of classes matches, same model architecture, same loss function, same color channels, same crop target size, etc. thoughts? some of these inconsistencies could cause train failure and would be less obvious at that point

File: train_segmentation_model.py (hunk 4 above)
Comment (CarolinaFurtado) [reply train_segmentation_model.py:207]: Can you please clarify? the metadata from the pretrained model is already included in the metadata of the new model

Comment (rak5216) [reply train_segmentation_model.py:207]: i see the pretrained config (pretrain input) but not the pretrained metadata (pretrain output)

File: train_segmentation_model.py (hunk 5 above)
Comment (CarolinaFurtado) [reply train_segmentation_model.py:251]: done

File: train_segmentation_model.py (hunk 6 above)
Comment (CarolinaFurtado) [reply train_segmentation_model.py:71]: done

File: train_segmentation_model.py (hunk 8 above)
Comment (CarolinaFurtado) [reply train_segmentation_model.py:79]: done

File: requirements.txt (hunk 3 above)
Comment (CarolinaFurtado) [reply requirements.txt:2]: true, done

File: configs/config_sandbox/tf_version_debug/train-small-3class_tfversion_debug.yaml (hunk 2 above)
Comment (CarolinaFurtado) [reply configs/config_sandbox/tf_version_debug/train-small-3class_tfversion_debug.yaml:8]: can I just remove the imagenet commented part?

File: configs/config_sandbox/tf_version_debug/dataset-small-3class_tfversion_debug.yaml (hunk 1 above)
Comment (CarolinaFurtado) [reply configs/config_sandbox/tf_version_debug/dataset-small-3class_tfversion_debug.yaml:21]: ok, I didn't know that, done

File: train_segmentation_model.py (hunk 4 above)
Comment (CarolinaFurtado) [reply train_segmentation_model.py:207]: like this? 
image

File: configs/config_sandbox/tf_version_debug/dataset-small-3class_tfversion_debug.yaml (hunk 1 above)
Comment (rak5216) [reply configs/config_sandbox/tf_version_debug/dataset-small-3class_tfversion_debug.yaml:21]: yeah, if u look into `prepare_dataset.py`, turns out we're just doing string comparisons with `=='None'` for type, but if a relevant type is selected, then a number needs to appear here. if linear or random types are not selected, then this value doesn't even matter

File: configs/config_sandbox/tf_version_debug/train-small-3class_tfversion_debug.yaml (hunk 2 above)
Comment (rak5216) [reply configs/config_sandbox/tf_version_debug/train-small-3class_tfversion_debug.yaml:8]: i was thinking about keeping it in case someone in the future has RGB datasets, so it wouldn't be forgotten. tho that's not our problem i guess, as the source can always be traced. the options also may change here once we allow additional architectures

File: train_segmentation_model.py (hunk 4 above)
Comment (rak5216) [reply train_segmentation_model.py:207]: yup that's awesome. just convert `pre_train` to `pretrain` then. 

also, for readability of the new model metadata, i would suggest we group all pretrained stuff under one dict key in the current model metadata dictionary (kind of like in train seg thresholds), which requires just a quick rewrite to form in advance the pretrained model id, config, and metadata into a separate dictionary first, defined like

```
 pretrained_model_info={'pretrained_id': pretrained_id, 'pretrained_config': pretrained_config, 'pretrained_meta': pretrained_meta}
```

and then current model metadata is just 

```
metadata = {
    the usual stuff...,
    'pretrained_model_info' : pretrained_model_info
}
```
or however u want to label things. since with current way, unordered current metadata dict could shuffle the pretrained content

Comment (CarolinaFurtado) [reply train_segmentation_model.py:207]: image
 done!

File: train_segmentation_model.py (hunk 9 above)
Comment (CarolinaFurtado) [reply train_segmentation_model.py]: any thoughts, @Josh-Joseph ?

File: train_segmentation_model.py (hunk 4 above)
Comment (rak5216) [reply train_segmentation_model.py]: looks great, last thing, didn't mean to suggest the labelling exactly, so u can keep `pretrained_model_metadata` or just `pretrained_metadata` (vs. `meta`), same for config and id. besides this, solid! u can imagine how this will nest nicely now if we pretrain a model an arbitrary number of times

File: train_segmentation_model.py (hunk 9 above)
Comment (Josh-Joseph) [reply train_segmentation_model.py]: yeah I think this is a great idea. @rak5216 I think you covered it -- generally I'd include asserts for:
- pretrained model name (is helpful to know if you, for example, misspelled something)
- input sizes
- architectures and pretrained backbones that we know don't play nicely together (you can imagine 6 months from now us completely forgetting which don't work and it'd be nice for an assert to tell us that)

**Review by Josh-Joseph — APPROVED** [review]
code looks good - no comments from me

File: train_segmentation_model.py (hunk 9 above)
Comment (CarolinaFurtado) [reply train_segmentation_model.py]: I created the asserts: 
image


E.g if the backbone of the trained and pretrained models are different, it fails (added the prints to make sure it was doing what I expected,  going to remove them now): 
image

if everything is equal, it works: 
image

Anyhing else missing @rak5216 @Josh-Joseph ?

File: train_segmentation_model.py [hunk 10]
```diff
@@ -89,17 +89,50 @@ def train(gcp_bucket, config_file, random_module_global_seed, numpy_random_globa
...
+        assert pretrained_model_config['segmentation_model']['model_parameters']['backbone_name'] == train_config['segmentation_model']['model_parameters']['backbone_name']
+        # same loss function
+        print(pretrained_model_config['loss'], train_config['loss'])
+        assert pretrained_model_config['loss'] == train_config['loss']
+        # confirm that the number of classes in pretrain is the same as train
+        print(pretrained_model_metadata['num_classes'], len(train_generator.mask_filenames))
+        assert pretrained_model_metadata['num_classes'] == len(train_generator.mask_filenames)
+        # same target size
+        print(pretrained_model_metadata['dataset_config']['target_size'], dataset_config['target_size'])
+        assert pretrained_model_metadata['dataset_config']['target_size'] == dataset_config['target_size']
```
Comment (rak5216) [inline train_segmentation_model.py:123]: last thing i can think of is to assert number of color channels (rgb vs. gv). somewhere in the code there is a hint how to check this

Comment (rak5216) [reply train_segmentation_model.py:123]: first layer of model would tell this, but accessing these layer dimensions may not be straightforward, so can skip for now. a model on gv will break with rgb. currently it's hardcoded for only GV (`model = Unet(input_shape=(None, None, 1),` in `models.py`), so we'll revisit this assertion (hopefully we remember) if/when we open up the code to accept rgb (via train_config)

File: train_segmentation_model.py [hunk 11]
```diff
@@ -89,17 +89,50 @@ def train(gcp_bucket, config_file, random_module_global_seed, numpy_random_globa
...
+            pretrained_model_metadata = yaml.safe_load(f)
+
+        pretrained_info = {'pretrained_model_id': pretrained_model_id, 'pretrained_config': pretrained_model_config, 'pretrained_metadata': pretrained_model_metadata}
+
+        # confirm that the same model and backbone are used from pretrain and train
+        print(pretrained_model_config['segmentation_model']['model_name'], train_config['segmentation_model']['model_name'])
+        assert pretrained_model_config['segmentation_model']['model_name'] == train_config['segmentation_model']['model_name']
+        print(pretrained_model_config['segmentation_model']['model_parameters']['backbone_name'], train_config['segmentation_model']['model_parameters']['backbone_name'])
+        assert pretrained_model_config['segmentation_model']['model_parameters']['backbone_name'] == train_config['segmentation_model']['model_parameters']['backbone_name']
+        # same loss function
```
[the remark this answers is not in the document]
Comment (rak5216) [reply train_segmentation_model.py:115]: this might be another instance to revisit. again, easiest way for now is actually just assert it like u r (ignore my prior comment). the real issue comes down to asserting binary loss functions (sigmoid mapped predictions) vs. categorical loss functions (softmax mapped predictions) rather than any one specific function. for example, there are variants of cross entropy, dice loss, jaccard loss, etc. currently we don't allow activation selection (sigmoid vs softmax) in train config, tho @BrendenBarbour may add this in his alt. models PR. here is an example of where activation is defaulted to currently (https://github.com/qubvel/segmentation_models/blob/master/segmentation_models/models/unet.py#L116).

long story short, leave loss fcn asserted, and we'll revisit with more nuance in brenden's PR i think

**Review by rak5216 — APPROVED** [review]
